src/System.Data.SqlClient/tests/FunctionalTests/DiagnosticTest.cs

```diff
@@ -0,0 +1,507 @@
+                {
+                    Assert.NotNull(kvp.Value);
+
+                    Guid retrievedOperationId = GetPropertyValueFromType<Guid>(kvp.Value, "OperationId");
+                    Assert.NotEqual(retrievedOperationId, Guid.Empty);
+
+                    SqlCommand sqlCommand = GetPropertyValueFromType<SqlCommand>(kvp.Value, "Command");
+                    Assert.NotNull(sqlCommand);
+
+                    string operation = GetPropertyValueFromType<string>(kvp.Value, "Operation");
+                    Assert.False(string.IsNullOrWhiteSpace(operation));
+
+                    beginOperationId = retrievedOperationId;
+
+                    statsLogged = true;
+                }
+                else if (kvp.Key.Equals("System.Data.SqlClient.WriteCommandAfter"))
+                {
+                    Assert.NotNull(kvp.Value);
+
+                    Guid retrievedOperationId = GetPropertyValueFromType<Guid>(kvp.Value, "OperationId");
+                    Assert.NotEqual(retrievedOperationId, Guid.Empty);
+
+                    SqlCommand sqlCommand = GetPropertyValueFromType<SqlCommand>(kvp.Value, "Command");
+                    Assert.NotNull(sqlCommand);
+
+                    statistics = GetPropertyValueFromType<IDictionary>(kvp.Value, "Statistics");
+                    if (!operationHasError)
+                        Assert.NotNull(statistics);
+
+                    string operation = GetPropertyValueFromType<string>(kvp.Value, "Operation");
+                    Assert.False(string.IsNullOrWhiteSpace(operation));
+
+                    // if we get to this point, then statistics exist and this must be the "end" 
+                    // event, so we need to make sure the operation IDs match
+                    Assert.Equal(retrievedOperationId, beginOperationId);
+                    beginOperationId = Guid.Empty;
+
+                    statsLogged = true;
+                }
+                else if (kvp.Key.Equals("System.Data.SqlClient.WriteCommandError"))
+                {
+                    operationHasError = true;
+                    Assert.NotNull(kvp.Value);
+
+                    SqlCommand sqlCommand = GetPropertyValueFromType<SqlCommand>(kvp.Value, "Command");
+                    Assert.NotNull(sqlCommand);
+
+                    Exception ex = GetPropertyValueFromType<Exception>(kvp.Value, "Exception");
+                    Assert.NotNull(ex);
+
+                    string operation = GetPropertyValueFromType<string>(kvp.Value, "Operation");
+                    Assert.False(string.IsNullOrWhiteSpace(operation));
+
+                    statsLogged = true;
+                }
+            });
+
+            diagnosticListenerObserver.Enable();
+            DiagnosticListener.AllListeners.Subscribe(diagnosticListenerObserver);
+
+            await sqlOperation();
+
+            Assert.True(statsLogged);
+
+            diagnosticListenerObserver.Disable();
+        }
+
+        private void ExecuteScalar()
```

Comment:
@stephentoub that's a good suggestion, I can refactor these tests with your recommendation.  Thanks! :smile: